### ISSUE TITLE
[corgan-pr-prep] T3: Refactor CorGAN to full PyHealth 2.0 BaseModel compliance

### ISSUE THREAD

**Issue by jalengg** [jalengg/PyHealth#28]
## Background

The current `pyhealth/models/generators/corgan.py` (commit 090caa4b) partially inherits `BaseModel` but uses non-standard API:
- `fit()` instead of `train_model()`
- `generate()` instead of `synthesize_dataset()`
- `_build_global_vocab()` instead of `dataset.input_processors["visits"].vocab_size()`
- `feature_keys`, `label_key`, `mode` params in `__init__` (not needed for generative model)

## Spec

1. **`__init__(self, dataset, ...)`**
   - Remove `feature_keys`, `label_key`, `mode` params
   - Get vocab from `dataset.input_processors["visits"].vocab_size()`
   - Remove `_build_global_vocab` method

2. **`train_model(self, train_dataset, val_dataset=None)`**
   - Rename from `fit()`
   - Accept `SampleDataset` (not raw DataLoader)
   - Build multi-hot encodings internally from nested sequence data
   - Save checkpoint after training

3. **`synthesize_dataset(self, num_samples, random_sampling=True) → List[Dict]`**
   - Rename from `generate()`
   - Return `[{"patient_id": "synthetic_N", "visits": [[codes...]]}]`
   - Each synthetic patient gets a SINGLE visit (CorGAN flattens temporal structure by design)
   - Decode multi-hot vectors back to ICD code strings using vocab

4. Keep all internal CNN/GAN components intact (`CorGANAutoencoder`, `CorGANGenerator`, `CorGANDiscriminator`, etc.)

## Key design note

CorGAN operates on per-patient multi-hot vectors (all visits aggregated into one binary vector). This is different from HALO's sequential visit model. Synthetic output is therefore a single-visit patient record — this is honest about the model's design, not a limitation of our integration.

## Acceptance criteria

- `train_model()` accepts `SampleDataset`, builds multi-hot internally
- `synthesize_dataset()` returns `List[Dict]` with correct schema
- No `_build_global_vocab` in final code
- All internal GAN components unchanged
- Spec compliance review ✅
- Code quality review ✅

**Comment by jalengg** [comment]
## Why this is the most complex task in the CorGAN integration

T3 is not a rename-and-done refactor. The current CorGAN model was written to run standalone — it manages its own vocabulary, its own data loading, and its own training loop in ways that bypass the PyHealth 2.0 abstractions entirely. Bringing it into compliance requires understanding what each current component does and finding the right PyHealth 2.0 equivalent, without breaking the GAN architecture underneath.

## The vocabulary problem: _build_global_vocab vs input_processors

The current code builds a global vocabulary by iterating over the entire dataset and collecting all unique ICD codes into a list. It then uses a `Tokenizer` to map codes to integer indices, and uses those indices to construct multi-hot binary vectors. This is correct but isolated — the vocabulary is stored inside the model instance and cannot be shared, inspected, or validated by the framework.

In PyHealth 2.0, when you call `dataset.set_task(corgan_generation_mimic3_fn)`, the framework automatically builds an `input_processor` for the `"visits"` field (a `NestedSequenceProcessor`). This processor constructs the vocabulary over the full training corpus and exposes it via `vocab_size()`. By the time `CorGAN.__init__` is called, the vocabulary already exists — `dataset.input_processors["visits"].vocab_size()` returns the exact count of unique codes including `<pad>` and `<unk>` tokens.

Replacing `_build_global_vocab` with `vocab_size()` is not just a code simplification. It guarantees that the vocabulary used during training is the same vocabulary used during generation — a subtle but critical contract that was not enforced in the original code (if you rebuilt the vocabulary on a different dataset split, you would get different code indices and the model would silently produce garbage).

## The method naming problem: fit() and generate()

The current CorGAN exposes `fit()` and `generate()`. PyHealth 2.0 requires `train_model()` and `synthesize_dataset()`.

The `fit` → `train_model` rename is non-negotiable because `fit` is a scikit-learn convention, not a PyHealth one. More importantly, `nn.Module.train()` is a PyTorch method that puts the model in training mode — naming a method `train` would shadow it and cause silent failures. `train_model` avoids this collision.

The `generate` → `synthesize_dataset` rename is about return type contract. The current `generate()` returns a raw `torch.Tensor` of multi-hot vectors. That is internal implementation detail, not a public API. `synthesize_dataset()` must return `List[Dict]` with `patient_id` and `visits` keys — the same format as real patient records from `set_task()`. This makes synthetic data directly comparable to real data and usable in downstream PyHealth evaluation pipelines.

## The critical design decision: single-visit synthetic patients

CorGAN operates on *patient-level* multi-hot encodings — it aggregates all ICD codes across all visits for a patient into a single binary vector and trains on those vectors. This is fundamentally different from HALO, which models the temporal sequence of visits explicitly.

This means that when CorGAN generates a synthetic patient, it produces a flat set of ICD codes with no visit structure. We cannot reverse-engineer visit boundaries from a multi-hot vector.

The correct and honest choice is to package each synthetic patient as a **single-visit record**: `{"patient_id": "synthetic_N", "visits": [["D001", "D002", ...]]}`. This is not a limitation of our integration — it accurately reflects what the model actually generates. A multi-visit synthetic patient reconstructed by some arbitrary split heuristic would be misleading.

This design decision should be documented in the `synthesize_dataset` docstring and acknowledged in the generation example.

## What must NOT change

The GAN architecture itself — `CorGANAutoencoder`, `CorGAN8LayerAutoencoder`, `CorGANLinearAutoencoder`, `CorGANGenerator`, `CorGANDiscriminator`, `weights_init`, `autoencoder_loss`, `discriminator_accuracy` — must not be touched. These implement the core CorGAN algorithm from the paper. Only the *interface layer* (how data flows in and out of the model) changes. The mathematical core is preserved exactly.

## The internal CorGANDataset helper

The current code contains a `CorGANDataset(Dataset)` class that wraps numpy arrays for PyTorch DataLoader. This is an internal implementation detail of `train_model` and is fine to keep. It is not a public PyHealth dataset class and should not be exported. Do not confuse it with the bespoke dataset classes that T6 removes.

## Why this ordering matters (T2 before T3)

T3 depends on T2 in one subtle way: the `train_model()` signature accepts a `SampleDataset` produced by `set_task(corgan_generation_mimic3_fn)`. Knowing the exact schema of that dataset (`input_schema = {"visits": "nested_sequence"}`) tells us what `dataset.input_processors["visits"]` contains and what `vocab_size()` returns. With T2 done, we have a concrete spec for T3's input type.

**Comment by jalengg** [comment]
## Schema change impact on T3: _encode_samples_to_multihot removed

Following the decision documented in issue #26 to use `multi_hot` instead of `nested_sequence`, the T3 model refactoring became significantly simpler.

### What was removed

`_encode_samples_to_multihot` — a 44-line method that iterated over the dataset, flattened visit lists, and manually looked up vocabulary indices to build a binary matrix — is **gone entirely**. It was the most complex part of the T3 refactoring, and it existed only because `nested_sequence` returned integer indices that still needed to be converted to binary vectors.

### What replaced it

In `train_model`, the data matrix is now built in two lines:

```python
tensors = [train_dataset[i]["visits"] for i in range(len(train_dataset))]
data_matrix = torch.stack(tensors).numpy()
```

`MultiHotProcessor` already produced float32 binary tensors of shape `(vocab_size,)` for each sample. `torch.stack` assembles them into the `(n_patients, vocab_size)` matrix that `CorGANDataset` and the training loop expect. No manual encoding, no custom collate function.

### Vocab size and decode

- `__init__`: `dataset.input_processors["visits"].size()` returns `vocab_size` (MultiHotProcessor uses `size()`, not `vocab_size()`)
- `synthesize_dataset`: reverse vocab lookup uses `processor.label_vocab` (the attribute name on `MultiHotProcessor`) to map indices back to code strings
- Output is a flat list of decoded code strings per patient — `{"visits": ["ICD1", "ICD2"]}` — honest about CorGAN generating bag-of-codes, not visit sequences

### Net result

The T3 diff is ~50 lines smaller than originally planned. The model is cleaner, the encoding path is shorter, and the data flow from task function → processor → model → generation is now consistent end-to-end with no hidden transformation steps.

**Comment by jalengg** [comment]
## T3 Complete ✅

**What was refactored:**
- `CorGAN(BaseModel)` now calls `super().__init__(dataset)` correctly (no feature_keys/label_key/mode — generator pattern)
- `__init__` derives vocab size from `dataset.input_processors["visits"].size()`
- `train_model(train_dataset, val_dataset=None)` replaces old `fit()` method
- `synthesize_dataset(num_samples, random_sampling=True)` replaces old `generate()` method
- `import os` moved to module top (was deferred inside a method)
- `_encode_samples_to_multihot` **removed entirely** — no longer needed with `multi_hot` schema
- `save_model` / `load_model` have full Google-style docstrings

**`multi_hot` schema impact on T3:**
With the `multi_hot` schema, `train_dataset[i]["visits"]` is already a `(vocab_size,)` float32 tensor. Building the training matrix is simply:
```python
tensors = [train_dataset[i]["visits"] for i in range(len(train_dataset))]
data_matrix = torch.stack(tensors).numpy()
```
No encoding step, no padding, no custom collate function needed.

**`synthesize_dataset` returns** `List[Dict]` where each dict has a flat `"visits"` list of code strings, decoded via `processor.label_vocab` reverse lookup. Documented edge case: all-zeros generation falls back to empty list.

**Commits:** `41e02a45` (initial refactor), `c02f1584` (multi_hot schema switch), `77bfd406` (quality fixes), `f3b79951` (docstring updates)

**Reviews:** Spec ✅ (all 15 requirements) | Code quality ✅
